SwiftCompilerSources/Sources/Optimizer/ModulePasses/DiagnoseUnknownConstValues.swift

```diff
@@ -0,0 +1,103 @@
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SIL
+import AST
+
+/// Performs mandatory diagnostic pass for emitting errors for '@const' values which the compiler was not able to
+/// simplify/interpret/reduce to a symbolic value.
+///
+let diagnoseUnknownConstValues = ModulePass(name: "diagnose-unknown-const-values") {
+  (moduleContext: ModulePassContext) in
+  var constExprState: ConstExprFunctionState = .init()
+  defer { constExprState.deinitialize() }
+  
+  // Verify all const globals to be initialized with compile-time known values
+  verifyGlobals(moduleContext)
+  
+  // Verify @const lets appearing as local variables
+  verifyLocals(constExprState: &constExprState,
+               moduleContext)
+  
+  // For each function call, ensure arguments to @const parameters are all compile-time known values
+  verifyCallArguments(constExprState: &constExprState,
+                      moduleContext)
+  
+  // For each `@const` function, ensure it is fully evaluable/interpretable at compile time
+  verifyFunctions(moduleContext)
+}
+
+private func verifyGlobals(_ context: ModulePassContext) {
+  for gv in context.globalVariables where gv.isConst {
+    if gv.staticInitValue == nil {
+      context.diagnosticEngine.diagnose(gv.sourceLocation,
+                                        .require_const_initializer_for_const)
+    }
+  }
+}
+
+private func verifyLocals(constExprState: inout ConstExprFunctionState,
+                          _ context: ModulePassContext) {
+  for f in context.functions {
+    for i in f.instructions {
+      if let dbi = i as? DebugValueInst {
+        verifyLocal(debugValueInst: dbi, constExprState: &constExprState, in: context)
+      }
+    }
+  }
+}
+
+private func verifyLocal(debugValueInst: DebugValueInst,
+                         constExprState: inout ConstExprFunctionState,
+                         in context: ModulePassContext) {
+  guard let localVarDecl = debugValueInst.varDecl,
+        !(localVarDecl is ParamDecl),
+        localVarDecl.isConst else {
+    return
+  }
+  
+  if !constExprState.isConstantValue(debugValueInst.operand.value) {
+    context.diagnosticEngine.diagnose(debugValueInst.location.sourceLoc,
+                                      .require_const_initializer_for_const)
+  }
+}
+
+private func verifyCallArguments(constExprState: inout ConstExprFunctionState,
+                                 _ context: ModulePassContext) {
+  for f in context.functions {
+    for i in f.instructions {
+      // TODO: Consider closures (partial_apply)
+      if let apply = i as? FullApplySite {
+        verifyCallArguments(apply: apply, constExprState: &constExprState, in: context)
+      }
+    }
+  }
+}
+
+private func verifyCallArguments(apply: ApplySite,
```

Comment:
It makes sense to replace `ApplySite` -> `FullApplySite` here, too, because this function assumes that the number of arguments == number of parameter (which is only the case for full apply sites).